### PR TITLE
Support scheduled transactions for existing rest apis

### LINKED ISSUE #1468: Scheduled transactions support for existing REST APIs
**Problem**
The existing REST APIs need to support scheduled transactions.

**Solution**
As noted in the [design](https://github.com/hashgraph/hedera-mirror-node/blob/master/docs/design/scheduled-transactions.md#rest-api):
- Update `openapi.yml`
- Update `/api/v1/transactions/:id`
- Update `/api/v1/transactions/:id/stateproof`

**Alternatives**

**Additional Context**
Depends on #1464

### PR DESCRIPTION
**Pull request by xin-hedera**
**Detailed description**:

- Support `scheduled` in endpoint `/api/v1/transactions/:id`
- Support `scheduled` in endpoint `/api/v1/transactions/:id/stateproof`
- Add `scheduled` query param to `openapi.yml`
- Add test cases

**Which issue(s) this PR fixes**:
Fixes #1468 

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-rest/__tests__/integrationDbOps.js [hunk 1]
```diff
@@ -84,7 +84,7 @@ const instantiateDatabase = async function () {
     console.log(`Started dockerized PostgreSQL ${schemaConfigs.docker.imageName}:${schemaConfigs.docker.tagName}`);
   }
 
-  console.log(`sqlConnection will use jdbc:postgresql://${config.db.host}:${config.db.port}/${config.db.name}`);
+  console.log(`sqlConnection will use postgresql://${dbUser}:${dbPassword}@${config.db.host}:${config.db.port}/${config.db.name}`);
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/integrationDbOps.js:87]: for debug purpose, show the complete postgresql url with username and password

File: hedera-mirror-rest/__tests__/utilsFilters.test.js [hunk 2]
```diff
@@ -160,50 +197,63 @@ describe('utils formatComparator tests', () => {
...
+    utils.buildComparatorFilter(constants.filterKeys.LIMIT, '2.3'),
+    utils.buildComparatorFilter(constants.filterKeys.SEQUENCE_NUMBER, '3.4'),
+    utils.buildComparatorFilter(constants.filterKeys.SEQUENCE_NUMBER, '<:2'),
+    utils.buildComparatorFilter(constants.filterKeys.SEQUENCE_NUMBER, '<=:2'),
+    utils.buildComparatorFilter(constants.filterKeys.ACCOUNT_PUBLICKEY, '3c3d546321ff6f63d701d2ec5c2'),
+    utils.buildComparatorFilter(constants.filterKeys.ACCOUNT_BALANCE, '23456789012345678901234'),
+    utils.buildComparatorFilter(constants.filterKeys.SCHEDULED, 'invalid'),
+  ];
+
+  invalidFilters.forEach((filter) => {
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/utilsFilters.test.js:253]: the negative tests should run against individual invalid filter

File: hedera-mirror-rest/__tests__/specs/transactions-25-specific-id-repeated-valid-params-schedule-true.spec.json [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        "payerAccountId": "0.0.10",
+        "recipientAccountId": "0.0.9",
+        "amount": 10,
+        "nodeAccountId": "0.0.3",
+        "treasuryAccountId": "0.0.98",
+        "scheduled": true
+      }
+    ]
+  },
+  "url": "/api/v1/transactions/0.0.10-1234567890-000000001?scheduled=true&scheduled=false&scheduled=true",
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/specs/transactions-25-specific-id-repeated-valid-params-schedule-true.spec.json:73]: Having it true, false, true makes it impossible to determine whether it was first in wins or last in wins.

File: hedera-mirror-rest/__tests__/specs/stateproof-13-success-result-repeated-scheduled-false.spec.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      "data/db/stateproof/address_book.sql",
+      "data/db/stateproof/address_book_entry.sql",
+      "data/db/stateproof/record_file.sql",
+      "data/db/stateproof/transaction.sql"
+    ]
+  },
+  "url": "/api/v1/transactions/0.0.94139-1570800762-443131000/stateproof?scheduled=true&scheduled=false&scheduled=false",
+  "responseStatus": 200,
+  "responseJson": {
+    "record_file": "c3d987c874294eb95b2df9fd36b0656623315769af2ef344c35868780102b5c0",
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/specs/stateproof-13-success-result-repeated-scheduled-false.spec.json:20]: We should consider adding the record file name and/or timestamp to this output in the future. Makes it hard to verify results or reconstruct the S3 object without that.

File: hedera-mirror-rest/__tests__/specs/transactions-02-specific-id.spec.json [hunk 5]
```diff
@@ -16,34 +16,99 @@
...
-        "consensus_timestamp": "1234567890000000001",
+        "consensus_timestamp": "1234567890000000003",
+        "valid_start_timestamp": "1234567890000000001",
         "payerAccountId": "0.0.10",
         "recipientAccountId": "0.0.9",
         "amount": 10,
         "nodeAccountId": "0.0.3",
-        "treasuryAccountId": "0.0.98"
+        "treasuryAccountId": "0.0.98",
+        "scheduled": true
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/specs/transactions-02-specific-id.spec.json:69]: It's really, really confusing that we are setting transaction fields like `scheduled` on a list of `cryptotransfers`. At the same time, we're also maintaining a list of `transactions` objects that don't have fields like `scheduled` but do contain a list of `transfers`. None of that makes any sense and is unnecessarily complicated. If these lists don't match the corresponding tables, it makes it very hard to grok. We should address that in the future.

File: hedera-mirror-rest/__tests__/specs/transactions-22-specific-id-scheduled-false.spec.json [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            "amount": 10
+          },
+          {
+            "account": "0.0.98",
+            "amount": 1
+          }
+        ],
+        "valid_duration_seconds": "11",
+        "valid_start_timestamp": "1234567890.000000001"
+      }
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/specs/transactions-22-specific-id-scheduled-false.spec.json:100]: This should also return the duplicate transaction. If they want to filter that out, they should use the `result` query parameter.

File: hedera-mirror-rest/__tests__/specs/transactions-24-specific-id-repeated-valid-params-schedule-false.spec.json [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        "payerAccountId": "0.0.10",
+        "recipientAccountId": "0.0.9",
+        "amount": 10,
+        "nodeAccountId": "0.0.3",
+        "treasuryAccountId": "0.0.98",
+        "scheduled": true
+      }
+    ]
+  },
+  "url": "/api/v1/transactions/0.0.10-1234567890-000000001?scheduled=false&scheduled=true&scheduled=false",
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/specs/transactions-24-specific-id-repeated-valid-params-schedule-false.spec.json:73]: Having it false, true, false makes it impossible to determine whether it was first in wins or last in wins.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 8]
```diff
@@ -192,6 +199,13 @@ paths:
       operationId: getStateproofInfo
       parameters:
         - $ref: '#/components/parameters/transactionIdPathParam'
+        - name: scheduled
+          in: query
+          description: >-
+            Filter stateproof by the transaction scheduled flag. When true, return stateproof for the scheduled
+            transaction. When false, return stateproof for the non-scheduled transaction. The default is false.
+          schema:
+            type: boolean
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/api/v1/openapi.yml:208]: sure

File: hedera-mirror-rest/transactions.js [hunk 9]
```diff
@@ -425,16 +427,47 @@ const getTransactions = async (req, res) => {
...
+  }
+
+  let scheduled = scheduledValues;
+  if (Array.isArray(scheduledValues)) {
+    scheduled = scheduledValues[scheduledValues.length - 1];
+  }
+
+  if (!utils.parseBooleanValue(scheduled)) {
+    const scheduleCreateProtoId = await transactionTypes.get('SCHEDULECREATE');
+    return `t.type = ${scheduleCreateProtoId}`;
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js:452]: this follows the design doc:

- If false, return only the `ScheduleCreate` transaction

It was implemented to show transactions with `scheduled = false`, but then I changed to follow the design.

File: hedera-mirror-rest/__tests__/specs/transactions-25-specific-id-repeated-valid-params-schedule-true.spec.json (hunk 3 above)
Comment (xin-hedera) [reply hedera-mirror-rest/__tests__/specs/transactions-25-specific-id-repeated-valid-params-schedule-true.spec.json:73]: Will change it to `false, false, true`

File: hedera-mirror-rest/__tests__/specs/transactions-22-specific-id-scheduled-false.spec.json (hunk 6 above)
Comment (xin-hedera) [reply hedera-mirror-rest/__tests__/specs/transactions-22-specific-id-scheduled-false.spec.json:100]: same as my response how to implement `schedule=false`. this is the expected result if `false` should only return the `ScheduleCreate` transation

File: hedera-mirror-rest/__tests__/specs/transactions-24-specific-id-repeated-valid-params-schedule-false.spec.json (hunk 7 above)
Comment (xin-hedera) [reply hedera-mirror-rest/__tests__/specs/transactions-24-specific-id-repeated-valid-params-schedule-false.spec.json:73]: will change it to `true, true, false`

File: hedera-mirror-rest/transactions.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-rest/transactions.js:452]: Yes, that logic in the design is true if that transaction ID is linked to a `ScheduleCreate`. For other regular transactions, we should expect scheduled=false to still return the regular transaction. Perhaps you can update the design to make that a little clearer.

File: hedera-mirror-rest/__tests__/integrationDbOps.js (hunk 1 above)
Comment (ijungmann) [reply hedera-mirror-rest/__tests__/integrationDbOps.js:87]: I agree

**Review by ijungmann — CHANGES_REQUESTED** [review]
Looks good, haven't seen anything additional to add, just what Steven has already mentioned.

File: hedera-mirror-rest/__tests__/integrationDbOps.js (hunk 1 above)
Comment (xin-hedera) [reply hedera-mirror-rest/__tests__/integrationDbOps.js:87]: Yeah I saw the code scan complain. If there is an easy way to mark it as FP once and forever I'd want to keep the change.

File: hedera-mirror-rest/__tests__/specs/stateproof-13-success-result-repeated-scheduled-false.spec.json (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-rest/__tests__/specs/stateproof-13-success-result-repeated-scheduled-false.spec.json:20]: Probably all the file names (record & sig).

Also change the address_books to a list of timestamp and content objects.
Address book suggestion is for a future task not this one.

File: hedera-mirror-rest/stateproof.js [hunk 10]
```diff
@@ -34,15 +35,17 @@ const {FileDownloadError} = require('./errors/fileDownloadError');
...
-let getSuccessfulTransactionConsensusNs = async (transactionId) => {
+let getSuccessfulTransactionConsensusNs = async (transactionId, scheduled) => {
   const sqlParams = [transactionId.getEntityId().getEncodedId(), transactionId.getValidStartNs()];
   const sqlQuery = `SELECT consensus_ns
        FROM transaction
        WHERE payer_account_id = $1
          AND valid_start_ns = $2
-         AND result = 22`; // only the successful transaction
+         AND result = 22
+         AND scheduled = ${scheduled}`; // only the successful transaction
```
Comment (Nana-EC) [inline hedera-mirror-rest/stateproof.js:48]: nit: comment now on the wrong line

File: hedera-mirror-rest/stateproof.js [hunk 11]
```diff
@@ -231,15 +234,32 @@ let downloadRecordStreamFilesFromObjectStorage = async (...partialFilePaths) =>
  */
 let canReachConsensus = (actualCount, totalCount) => actualCount >= Math.ceil(totalCount / 3.0);
 
+/**
+ * Get scheduled param value from filters. In case there are multiple scheduled params in the filter, the last one is
+ * used. The default is false if not present.
+ * @param filters - the validated filters built from the HTTP request query
+ * @return {Boolean}
+ */
+const getScheduledParamValue = (filters) => {
```
Comment (Nana-EC) [inline hedera-mirror-rest/stateproof.js:243]: I'm sure this logic is useful for other params where we want last in logic.
We should consider moving this logic into the `utils.buildFilterObject()`
Whereby there's a set of params that don't support repetition. If encounter they don't create an array but just overwrite the previous value

File: hedera-mirror-rest/utils.js [hunk 12]
```diff
@@ -156,6 +160,10 @@ const filterValidityChecks = async (param, op, val) => {
       // Acceptable forms: exactly 64 characters or +12 bytes (DER encoded)
       ret = isValidPublicKeyQuery(val);
       break;
+    case constants.filterKeys.CREDIT_TYPE:
+      // Acceptable words: credit or debit
+      ret = Object.values(constants.cryptoTransferType).includes(val.toLowerCase());
```
Comment (Nana-EC) [inline hedera-mirror-rest/utils.js:165]: nit: convention would be to move this to a `isValidCreditType()`

**Review by Nana-EC — CHANGES_REQUESTED** [review]

**Review by steven-sheehy — DISMISSED** [review]
LGTM

File: hedera-mirror-rest/stateproof.js (hunk 10 above)
Comment (xin-hedera) [reply hedera-mirror-rest/stateproof.js:48]: can't have the comment in the template literal, but I can swap `result = 22` and `scheduled = ${scheduled}`

File: hedera-mirror-rest/utils.js (hunk 12 above)
Comment (xin-hedera) [reply hedera-mirror-rest/utils.js]: added a general function `isValidValueIgnoreCase(value, validValues)`

File: hedera-mirror-rest/stateproof.js (hunk 10 above)
Comment (Nana-EC) [reply hedera-mirror-rest/stateproof.js]: Haha you're right. Ignore that then

**Review by Nana-EC — APPROVED** [review]
Looks good.
My comment about moving the single value logic of `getScheduledParamValue()` to a shared utils file can be addressed in a separate PR if it's beyond this scope.

**Review by ijungmann — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**